PDF /Rotate metadata is not honored: pdf_info returns MediaBox dims and rasters/tiles do not follow page rotation

## Summary

`libviprs` does not consistently honor a PDF page's `/Rotate` metadata.
A page declared as `/Rotate 90` (clockwise) is reported and rendered using its raw `MediaBox` dimensions rather than the displayed (rotated) dimensions. Downstream tile pyramids therefore retain the source's *unrotated* orientation — e.g. a landscape `MediaBox` with `/Rotate 90` (which any PDF viewer shows as portrait) still produces landscape rasters and landscape tile grids.

This causes two visible symptoms downstream:

1. The reported page dimensions (`PdfPageInfo::{width_pts, height_pts}`) disagree with how the page is displayed in PDF viewers (and with what consumers reasonably expect "page width/height" to mean).
2. Tile pyramids generated from a rotated page do not adjust to the displayed orientation — a viewer-portrait page comes out as a landscape tile grid.

The expected behavior is: `pdf_info` and `render_page_pdfium*` should both report and render the page in its **displayed** orientation, with `/Rotate` already applied. Consumers that need the raw `MediaBox` orientation can ask for it explicitly.

## Affected APIs

- `pdf::pdf_info` → `PdfPageInfo::{width_pts, height_pts}` (file: `src/pdf.rs`, fn `get_page_dimensions`, ~L425)
  Reads `MediaBox` via `lopdf` and ignores the page's `/Rotate` entry.
- `pdf::render_page_pdfium` (`src/pdf.rs`, ~L602)
- `pdf::render_page_pdfium_budgeted` (`src/pdf.rs`, ~L657)
  Compute target raster size from `pdf_page.width()` / `pdf_page.height()` but call `render_at_size` with a `PdfRenderConfig` whose `portrait_rotation` / `landscape_rotation` are `None` — the rendered raster ends up in `MediaBox` orientation regardless of `/Rotate`.

## Reproduction

The reproduction below builds a tiny in-memory PDF with `MediaBox = [0, 0, 1000, 500]` (landscape) and `/Rotate 90`. A correctly-behaving library should report it as `500 × 1000` (portrait) and produce a portrait raster.

### Drop-in test suite

Add `src/pdf_rotation_tests.rs` (or extend the existing `#[cfg(test)] mod tests` in `src/pdf.rs`) with the suite below, plus a small in-memory PDF builder to keep the tests hermetic (no fixture files).

```rust
//! Tests for /Rotate handling — see GitHub issue
//! "PDF /Rotate metadata is not honored".
//!
//! Each test builds a minimal in-memory PDF with a known MediaBox and
//! /Rotate value, then exercises libviprs's public surface and asserts
//! the *displayed* (post-rotation) orientation.

use std::io::Write;
use tempfile::NamedTempFile;

use libviprs::{pdf_info, render_page_pdfium, render_page_pdfium_budgeted};

/// Build a 1-page PDF with the given MediaBox and /Rotate, write it to a
/// temp file, return the path. Uses lopdf so the test has zero external
/// fixture files.
fn build_pdf(media_box: [f64; 4], rotate: i64) -> NamedTempFile {
    use lopdf::{dictionary, Document, Object, Stream};

    let mut doc = Document::with_version("1.5");
    let pages_id = doc.new_object_id();
    let content = Stream::new(dictionary! {}, b"q 0.9 0.9 0.9 rg 0 0 100 100 re f Q".to_vec());
    let content_id = doc.add_object(content);

    let page_id = doc.add_object(dictionary! {
        "Type" => "Page",
        "Parent" => pages_id,
        "MediaBox" => vec![
            media_box[0].into(), media_box[1].into(),
            media_box[2].into(), media_box[3].into(),
        ],
        "Rotate" => rotate,
        "Contents" => content_id,
        "Resources" => dictionary! {},
    });

    doc.objects.insert(pages_id, Object::Dictionary(dictionary! {
        "Type" => "Pages",
        "Kids" => vec![page_id.into()],
        "Count" => 1,
    }));
    let catalog_id = doc.add_object(dictionary! {
        "Type" => "Catalog",
        "Pages" => pages_id,
    });
    doc.trailer.set("Root", catalog_id);

    let mut tmp = NamedTempFile::new().expect("temp file");
    let mut buf: Vec<u8> = Vec::new();
    doc.save_to(&mut buf).expect("serialize pdf");
    tmp.write_all(&buf).expect("write pdf");
    tmp
}

/// pdf_info should return *displayed* dimensions: a landscape MediaBox
/// with /Rotate 90 must be reported as portrait.
#[test]
fn pdf_info_honors_rotate_90_clockwise() {
    let pdf = build_pdf([0.0, 0.0, 1000.0, 500.0], 90);
    let info = pdf_info(pdf.path()).expect("pdf_info");
    let p = &info.pages[0];

    // Currently fails: get_page_dimensions reads MediaBox only and
    // returns (1000.0, 500.0), ignoring /Rotate.
    assert_eq!(p.width_pts, 500.0, "width should be the rotated (displayed) width");
    assert_eq!(p.height_pts, 1000.0, "height should be the rotated (displayed) height");
}

/// /Rotate 0 must be the identity case.
#[test]
fn pdf_info_no_rotate_is_identity() {
    let pdf = build_pdf([0.0, 0.0, 1000.0, 500.0], 0);
    let info = pdf_info(pdf.path()).expect("pdf_info");
    let p = &info.pages[0];
    assert_eq!(p.width_pts, 1000.0);
    assert_eq!(p.height_pts, 500.0);
}

/// /Rotate 180 should NOT swap dimensions.
#[test]
fn pdf_info_rotate_180_preserves_dims() {
    let pdf = build_pdf([0.0, 0.0, 1000.0, 500.0], 180);
    let info = pdf_info(pdf.path()).expect("pdf_info");
    let p = &info.pages[0];
    assert_eq!(p.width_pts, 1000.0);
    assert_eq!(p.height_pts, 500.0);
}

/// /Rotate 270 should swap dimensions (same shape as 90).
#[test]
fn pdf_info_rotate_270_swaps_dims() {
    let pdf = build_pdf([0.0, 0.0, 1000.0, 500.0], 270);
    let info = pdf_info(pdf.path()).expect("pdf_info");
    let p = &info.pages[0];
    assert_eq!(p.width_pts, 500.0);
    assert_eq!(p.height_pts, 1000.0);
}

/// The rendered raster must have the displayed orientation, not the
/// raw MediaBox orientation.
#[test]
#[cfg(feature = "pdfium")]
fn render_page_pdfium_honors_rotate_90_clockwise() {
    let pdf = build_pdf([0.0, 0.0, 1000.0, 500.0], 90);
    let raster = render_page_pdfium(pdf.path(), 1, 72).expect("render");

    // After /Rotate 90, the displayed page is 500 wide x 1000 tall.
    // At 72 DPI, the raster should also be portrait (height > width).
    assert!(
        raster.height() > raster.width(),
        "expected portrait raster (h > w) for landscape MediaBox + /Rotate 90, got {}x{}",
        raster.width(), raster.height()
    );
}

/// pdf_info and render must agree on aspect ratio.
#[test]
#[cfg(feature = "pdfium")]
fn pdf_info_and_render_agree_on_orientation_for_rotated_page() {
    let pdf = build_pdf([0.0, 0.0, 1000.0, 500.0], 90);

    let info = pdf_info(pdf.path()).expect("pdf_info");
    let p = &info.pages[0];
    let info_portrait = p.height_pts > p.width_pts;

    let r = render_page_pdfium_budgeted(pdf.path(), 1, 72, 10_000_000).expect("render");
    let raster_portrait = r.raster.height() > r.raster.width();

    assert_eq!(
        info_portrait, raster_portrait,
        "pdf_info reports portrait={info_portrait} but raster is portrait={raster_portrait} \
         (info: {}x{}, raster: {}x{})",
        p.width_pts, p.height_pts, r.raster.width(), r.raster.height()
    );
}
```

`tempfile` and `lopdf` are already in the dependency tree, so the only additional `[dev-dependencies]` entry needed is `tempfile` (it may already be present transitively for other tests).

### Expected outcome

Today, on `main` (and on `0.1.4`), the `pdf_info_honors_rotate_*` tests fail because `get_page_dimensions` does not read `/Rotate` from the page dictionary. The `pdf_info_and_render_agree_on_orientation_for_rotated_page` test will also fail because the two paths derive orientation from different sources (`MediaBox` for `pdf_info`, `pdf_page.width()`/`.height()` for the renderer, where the latter reflects whatever pdfium-render reports).

After the fix, all six tests should pass. They double as a regression suite — any future change that desyncs `pdf_info` and the renderer with respect to page orientation will trip the round-trip test.

## Suggested fix

1. In `get_page_dimensions` (`src/pdf.rs:425`), after reading `MediaBox`, also resolve the page's `/Rotate` entry (inheriting from the parent `Pages` tree per PDF 1.7 §7.7.3.3). When `/Rotate` is `90` or `270`, swap the returned `(width, height)`.
2. In `render_at_size` / the budgeted path, decide on a single source of truth for the rotated dimensions and ensure the `PdfRenderConfig` is configured (or the `PdfPage::set_rotation` call is made) so the raster matches the displayed orientation. The simplest option is to call `pdf_page.set_rotation(PdfPageRenderRotation::None)` and apply rotation explicitly in the config based on the value read in step 1 — that decouples libviprs from any future change in pdfium-render's default behavior.
3. Add the test suite above to lock the contract in.

Happy to send a PR if helpful — wanted to file the bug + repro first so the expected contract is agreed before the change lands.

## Environment

- `libviprs` 0.1.2 / 0.1.4 (verified both)
- `pdfium-render` 0.8.37
- Rust 1.x stable, `pdfium-static` feature enabled
- macOS (Darwin 25.x) and Linux (Debian bookworm) — same behavior on both

## Follow-up: restoring true strip rendering for pdfium sources (Option B — fork `pdfium-render`)

### Why this is a follow-up

PR #51 fixes correctness by rendering the full display-oriented page in `PdfiumStripSource::new`/`new_with_budget` and slicing strips from the cache. Correctness-first is the right call, but it regresses source-side peak memory to `O(canvas_w × canvas_h)` — the advantage the strip engine was designed around is gone for pdfium sources until we follow up.

For context on the memory cost:
- `blueprint.pdf` at 72 DPI → 4768×3370 RGBA → ~64 MB per `PdfiumStripSource`.
- Same page at 300 DPI → ~1.1 GB.

The downstream streaming pipeline's per-strip profile is unchanged — tiles still flow at the same cadence — but the source-side memory ceiling is back to monolithic.

### What pdfium's C API can do

`FPDF_RenderPageBitmap(bitmap, page, start_x, start_y, size_x, size_y, rotate, flags)` supports a "window-onto-page" mode that gives us both strip memory and correct `/Rotate`:

- Allocate a **strip-sized** bitmap `(full_w, strip_h)`.
- Pass `size_x = full_w`, `size_y = full_h` — i.e. tell pdfium it's rendering the *full* page.
- Pass `start_x = 0`, `start_y = -y_offset` — negative, shifts the page up so only rows `[y_offset .. y_offset + strip_h)` land in the bitmap.
- Pass `rotate = 0`. pdfium still applies the page's intrinsic `/Rotate` on this path; the `rotate` argument is an *additional* caller-requested rotation on top.

Memory: `O(full_w × strip_h)` on the source side. Rotation: correct for free. No matrix math. This is the approach libvips uses for pdfium-backed tile rendering.

### Why the Rust code can't do this today

The `pdfium-render` crate (v0.8.37) has two blockers:

1. `PdfPage::render_into_bitmap_with_settings` (`src/pdf/document/page.rs:655-664`) calls `FPDF_RenderPageBitmap` with **`start_x = 0, start_y = 0` hardcoded** when the form-data path is taken. The form-data path is the one that auto-applies `/Rotate`, so any caller that wants both `/Rotate` correctness and a strip-sized bitmap is locked out.
2. `PdfPage::page_handle` and `PdfPage::bindings` are `pub(crate)`, so callers can't bypass the wrapper and invoke `FPDF_RenderPageBitmap` directly either.

Verified empirically — a diagnostic that tries to call the raw bindings from outside the crate fails to compile:

```
error[E0624]: method `page_handle` is private
   --> main.rs:49:14
    |
49  |         page.page_handle(),
    |              ^^^^^^^^^^^ private method
```

### Proposed fix — Option B

Fork `pdfium-render` into the `libviprs` GitHub organization. Apply a minimal patch that exposes the window parameters — either:

- add a new `PdfPage::render_window_into_bitmap(bitmap, start_x, start_y, size_x, size_y, rotate, flags)` method that dispatches straight to `FPDF_RenderPageBitmap`, or
- accept optional `start_x` / `start_y` on the existing `PdfRenderConfig` builder and thread them through `apply_to_page` → `render_into_bitmap_with_settings`.

Pin via `[patch.crates-io]` in the workspace `Cargo.toml`. Submit the same patch upstream as a PR in parallel; when/if it merges and a release ships, drop the `[patch.crates-io]` entry and the fork.

### Downstream work once the patch lands

1. `PdfiumStripSource` constructor stops materialising the full raster. No cached `Raster` field; dimensions come from a cheap probe (can reuse the existing `probe_page_dims` since the rounding drift only mattered when the raster was sliced).
2. `render_strip_inner` allocates a strip-sized BGRA bitmap, calls the new window API with `start_y = -y_offset`, `size_x = full_w`, `size_y = full_h`, `rotate = 0`, flags `0`.
3. Re-run the parity diagnostic from #52 (raw strip vs. full render → byte-identical) and the pyramid-level monolithic-vs-streaming comparison. Expectation: still byte-identical, now at `O(canvas_w × strip_h)` source memory.
4. Unignore / restore any streaming tests that currently rely on the cached full raster behavior.

### Risk and surface area

- Scope is confined to `src/streaming.rs` (`PdfiumStripSource`) plus the one-line `[patch.crates-io]` entry and the fork.
- The fork is a trivial additive change against a tagged `pdfium-render` release; easy to rebase onto newer releases.
- No API break on `libviprs` — `StripSource::width`/`height`/`render_strip` contracts are unchanged. The change is purely an implementation swap behind `PdfiumStripSource`.
- Parity tests from PR #51 act as regression gates for the swap.

--- a/src/pdf.rs
+++ b/src/pdf.rs
@@ -421,7 +421,14 @@ fn cmyk_to_rgb_raster(cmyk_data: &[u8], width: u32, height: u32) -> Result<Raste
     Raster::new(width, height, PixelFormat::Rgb8, rgb).map_err(PdfError::Raster)
 }
 
-/// Get page dimensions in points from MediaBox.
+/// Get page dimensions in points, in the *displayed* orientation.
+///
+/// Reads `/MediaBox` and applies the effective `/Rotate` from the page
+/// dictionary (inheriting through the `/Parent` chain per PDF 1.7
+/// §7.7.3.3). For `/Rotate 90` or `270`, width and height are swapped so
+/// the returned `(w, h)` matches what viewers and the pdfium form-data
+/// render path report. `/Rotate 0` and `180` preserve the MediaBox
+/// orientation. Missing `/Rotate` behaves as `0`.
 fn get_page_dimensions(doc: &lopdf::Document, page_id: lopdf::ObjectId) -> (f64, f64) {
     let obj = match doc.get_object(page_id) {
         Ok(o) => o,
@@ -439,11 +446,69 @@ fn get_page_dimensions(doc: &lopdf::Document, page_id: lopdf::ObjectId) -> (f64,
             let y0 = obj_to_f64(&media_box[1]).unwrap_or(0.0);
             let x1 = obj_to_f64(&media_box[2]).unwrap_or(0.0);
             let y1 = obj_to_f64(&media_box[3]).unwrap_or(0.0);
-            return ((x1 - x0).abs(), (y1 - y0).abs());
+            let w = (x1 - x0).abs();
+            let h = (y1 - y0).abs();
+            return apply_rotate_to_dims(w, h, resolve_rotate(doc, page_id));
         }
     }
 
     (0.0, 0.0)
+}
+
+/// Swap `(w, h)` when `rotate` is `90` or `270` (mod 360). Returns
+/// `(w, h)` unchanged for `0`, `180`, or any non-multiple-of-90 value.
+fn apply_rotate_to_dims(w: f64, h: f64, rotate: i64) -> (f64, f64) {
+    // PDF spec: /Rotate must be a multiple of 90. `rem_euclid` normalises
+    // negative or out-of-range values (seen in malformed PDFs) into
+    // `[0, 360)` before the swap decision.
+    let normalised = rotate.rem_euclid(360);
+    if normalised == 90 || normalised == 270 {
+        (h, w)
+    } else {
+        (w, h)
+    }
+}
+
+/// Resolve the effective `/Rotate` value for a page.
+///
+/// Walks the `/Parent` chain per PDF 1.7 §7.7.3.3: the page's own dict
+/// first, then each ancestor `Pages` node, returning the first numeric
+/// `/Rotate` encountered. Returns `0` if none is found or the traversal
+/// hits a malformed node / self-referential loop.
+fn resolve_rotate(doc: &lopdf::Document, page_id: lopdf::ObjectId) -> i64 {
+    let mut current = page_id;
+    // Cap the walk defensively — page trees more than a few dozen levels
+    // deep are pathological, and this stops an adversarial self-referential
+    // `/Parent` from spinning forever even if the ID-equality guard below
+    // misses it (e.g. alternating pairs of ids).
+    for _ in 0..64 {
+        let Ok(obj) = doc.get_object(current) else {
+            return 0;
+        };
+        let Ok(dict) = obj.as_dict() else {
+            return 0;
+        };
+        if let Ok(rotate_obj) = dict.get(b"Rotate") {
+            if let Ok(resolved) = resolve_object(doc, rotate_obj) {
+                if let Some(v) = obj_to_f64(resolved) {
+                    return v as i64;
+                }
+            }
+        }
+        // `/Parent` is an indirect reference to the parent Pages node.
+        // Pull the object id straight from the Reference — don't call
+        // `resolve_object`, which would hand back the dict itself and
+        // lose the id we need to walk upward.
+        let parent_id = match dict.get(b"Parent") {
+            Ok(lopdf::Object::Reference(id)) => *id,
+            _ => return 0,
+        };
+        if parent_id == current {
+            return 0;
+        }
+        current = parent_id;
+    }
+    0
 }
 
 /// Check whether a page has any Image XObjects.
@@ -765,5 +830,132 @@ mod tests {
     fn obj_to_f64_other() {
         let obj = lopdf::Object::Boolean(true);
         assert_eq!(obj_to_f64(&obj), None);
+    }
+
+    // -----------------------------------------------------------------
+    // /Rotate handling (issue #50)
+    //
+    // These tests pin the contract that `get_page_dimensions` returns
+    // display-oriented dimensions: a landscape MediaBox with /Rotate 90
+    // is reported as portrait, /Rotate 0 and /Rotate 180 preserve the
+    // raw MediaBox orientation, and /Rotate is inheritable through the
+    // Pages tree. Each test builds a minimal in-memory PDF via lopdf
+    // so the contract is exercised end-to-end through the same parser
+    // path `pdf_info` uses at runtime.
+    // -----------------------------------------------------------------
+
+    /// Build a minimal 1-page document with the given MediaBox +
+    /// `page_rotate` on the leaf page and `pages_rotate` on the parent
+    /// Pages node. `None` for either means the key is absent.
+    fn build_rotated_doc(
+        media_box: [f64; 4],
+        page_rotate: Option<i64>,
+        pages_rotate: Option<i64>,
+    ) -> (lopdf::Document, lopdf::ObjectId) {
+        use lopdf::{Document, Object, Stream, dictionary};
+
+        let mut doc = Document::with_version("1.5");
+        let pages_id = doc.new_object_id();
+        let content = Stream::new(dictionary! {}, b"q 0 g 0 0 10 10 re f Q".to_vec());
+        let content_id = doc.add_object(content);
+
+        let mut page_dict = dictionary! {
+            "Type" => "Page",
+            "Parent" => pages_id,
+            "MediaBox" => vec![
+                media_box[0].into(),
+                media_box[1].into(),
+                media_box[2].into(),
+                media_box[3].into(),
+            ],
+            "Contents" => content_id,
+            "Resources" => dictionary! {},
+        };
+        if let Some(r) = page_rotate {
+            page_dict.set("Rotate", Object::Integer(r));
+        }
+        let page_id = doc.add_object(page_dict);
+
+        let mut pages_dict = dictionary! {
+            "Type" => "Pages",
+            "Kids" => vec![page_id.into()],
+            "Count" => 1i64,
+        };
+        if let Some(r) = pages_rotate {
+            pages_dict.set("Rotate", Object::Integer(r));
+        }
+        doc.objects.insert(pages_id, Object::Dictionary(pages_dict));
+
+        let catalog_id = doc.add_object(dictionary! {
+            "Type" => "Catalog",
+            "Pages" => pages_id,
+        });
+        doc.trailer.set("Root", catalog_id);
+        (doc, page_id)
+    }
+
+    #[test]
+    fn get_page_dimensions_no_rotate_is_identity() {
+        let (doc, page_id) = build_rotated_doc([0.0, 0.0, 1000.0, 500.0], None, None);
+        assert_eq!(get_page_dimensions(&doc, page_id), (1000.0, 500.0));
+    }
+
+    #[test]
+    fn get_page_dimensions_rotate_zero_is_identity() {
+        let (doc, page_id) = build_rotated_doc([0.0, 0.0, 1000.0, 500.0], Some(0), None);
+        assert_eq!(get_page_dimensions(&doc, page_id), (1000.0, 500.0));
+    }
+
+    #[test]
+    fn get_page_dimensions_rotate_90_swaps() {
+        let (doc, page_id) = build_rotated_doc([0.0, 0.0, 1000.0, 500.0], Some(90), None);
+        assert_eq!(get_page_dimensions(&doc, page_id), (500.0, 1000.0));
+    }
+
+    #[test]
+    fn get_page_dimensions_rotate_180_preserves() {
+        let (doc, page_id) = build_rotated_doc([0.0, 0.0, 1000.0, 500.0], Some(180), None);
+        assert_eq!(get_page_dimensions(&doc, page_id), (1000.0, 500.0));
+    }
+
+    #[test]
+    fn get_page_dimensions_rotate_270_swaps() {
+        let (doc, page_id) = build_rotated_doc([0.0, 0.0, 1000.0, 500.0], Some(270), None);
+        assert_eq!(get_page_dimensions(&doc, page_id), (500.0, 1000.0));
+    }
+
+    #[test]
+    fn get_page_dimensions_negative_rotate_normalises() {
+        // /Rotate -90 is equivalent to 270 under rem_euclid normalisation.
+        let (doc, page_id) = build_rotated_doc([0.0, 0.0, 1000.0, 500.0], Some(-90), None);
+        assert_eq!(get_page_dimensions(&doc, page_id), (500.0, 1000.0));
+    }
+
+    #[test]
+    fn get_page_dimensions_rotate_inherited_from_parent_pages() {
+        // /Rotate on the parent Pages node, absent on the page itself.
+        // PDF 1.7 §7.7.3.3 mandates inheritance; the page must pick up
+        // the parent's rotation.
+        let (doc, page_id) = build_rotated_doc([0.0, 0.0, 1000.0, 500.0], None, Some(90));
+        assert_eq!(get_page_dimensions(&doc, page_id), (500.0, 1000.0));
+    }
+
+    #[test]
+    fn get_page_dimensions_page_rotate_overrides_parent() {
+        // When both the page and the parent Pages node declare /Rotate,
+        // the page's value wins — the walk stops at the first /Rotate it
+        // encounters.
+        let (doc, page_id) = build_rotated_doc([0.0, 0.0, 1000.0, 500.0], Some(0), Some(90));
+        assert_eq!(get_page_dimensions(&doc, page_id), (1000.0, 500.0));
+    }
+
+    #[test]
+    fn apply_rotate_to_dims_multiples() {
+        assert_eq!(apply_rotate_to_dims(1000.0, 500.0, 0), (1000.0, 500.0));
+        assert_eq!(apply_rotate_to_dims(1000.0, 500.0, 90), (500.0, 1000.0));
+        assert_eq!(apply_rotate_to_dims(1000.0, 500.0, 180), (1000.0, 500.0));
+        assert_eq!(apply_rotate_to_dims(1000.0, 500.0, 270), (500.0, 1000.0));
+        assert_eq!(apply_rotate_to_dims(1000.0, 500.0, 360), (1000.0, 500.0));
+        assert_eq!(apply_rotate_to_dims(1000.0, 500.0, -90), (500.0, 1000.0));
     }
 }

